python/ray/serve/_private/router.py

```diff
@@ -83,62 +141,248 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
+
+        After each iteration, there will be an increasing backoff sleep time (dictated
+        by `self.backoff_sequence_s`). The caller should exit the generator to reset the
+        backoff sleep time.
+        """
+        backoff_index = 0
+        while True:
+            # If no replicas are available, wait until `update_replicas` is called.
+            while len(self._replicas) == 0:
                 logger.info(
                     "Tried to assign replica for deployment "
-                    f"{self._deployment_name} but none available.",
+                    f"{self._deployment_name} but none are available.",
                     extra={"log_to_stderr": False},
                 )
-                # Clear before waiting to avoid immediately waking up if there
-                # had ever been an update before.
                 self._replicas_updated_event.clear()
                 await self._replicas_updated_event.wait()
+                logger.info(
+                    "Got replicas for deployment {self._deployment_name}, waking up.",
+                    extra={"log_to_stderr": False},
+                )
 
-        if replica.is_cross_language:
-            raise RuntimeError(
-                "Streaming is not yet supported for cross-language actors."
+            chosen_ids = random.sample(
+                self._replica_id_set, k=min(2, len(self._replica_id_set))
             )
+            yield [self._replicas[chosen_id] for chosen_id in chosen_ids]
 
-        return replica.actor_handle.handle_request_streaming.options(
-            num_returns="streaming"
-        ).remote(
-            pickle.dumps(query.metadata),
-            *query.args,
-            **query.kwargs,
+            await asyncio.sleep(self.backoff_sequence_s[backoff_index])
+            backoff_index = min(backoff_index + 1, len(self.backoff_sequence_s) - 1)
+
+    async def select_from_candidate_replicas(
+        self, candidates: List[ReplicaWrapper], response_deadline_s: float = 0.1
+    ) -> Optional[ReplicaWrapper]:
+        """Chooses the best replica from the list of candidates.
+
+        If none of the replicas can be scheduled, returns `None`.
+
+        The queue length at each replica is queried directly from it. The time waited
+        for these queries is capped by `response_deadline_s`; if a replica doesn't
+        respond within the deadline it is not considered.
+
+        Among replicas that respond within the deadline and accept the request (don't
+        have full queues), the one with the lowest queue length is chosen.
+        """
+        get_queue_len_tasks = [c.get_queue_len() for c in candidates]
+        done, pending = await asyncio.wait(
+            get_queue_len_tasks,
+            timeout=response_deadline_s,
+            return_when=asyncio.ALL_COMPLETED,
         )
+        for task in pending:
+            task.cancel()
+
+        chosen_replica_id = None
+        lowest_queue_len = math.inf
```

Comment:
[Nit] Up to you, I prefer this because it removes an import, but I'm okay either way.

```suggestion
        lowest_queue_len = float("inf")
```